src/sink.jl

```diff
@@ -0,0 +1,120 @@
+              reference::Vector{UInt8}=UInt8[])
+    Sink(sink.filename, sch, sink.columns)
+end
+
+Data.streamtypes(::Type{Sink}) = [Data.Column]
+
+size(sink::Sink) = size(sink.schema)
+size(sink::Sink, i::Integer) = size(sink.schema, i)
+
+
+"""
+    write(filename::AbstractString, df::DataFrame)
+
+Write the dataframe `df` to the feather formatted file `filename`.
+"""
+function write(filename::AbstractString, df::AbstractDataFrame)
+    sink = Feather.Sink(filename, df)
+    Data.stream!(df, sink)
+    Data.close!(sink)
+end
+
+
+function Data.streamto!(sink::Sink, ::Type{Data.Column}, val::AbstractVector{T}, row, col) where T
+    sink.columns[col] = arrowformat(val)
+end
+
+# NOTE: the below is very inefficient, but we are forced to do it by the Feather format
+function Data.streamto!(sink::Sink, ::Type{Data.Column}, val::AbstractVector{Union{T,Missing}},
+                        row, col) where T
+    hasmissing = Compat.findfirst(ismissing, val)
+    sink.columns[col] = arrowformat(hasmissing == nothing ? convert(AbstractVector{T}, val) : val)
+end
+
+
+function Metadata.PrimitiveArray(A::ArrowVector{J}, off::Integer, nbytes::Integer) where J
+    Metadata.PrimitiveArray(feathertype(J), Metadata.PLAIN, off, length(A), nullcount(A), nbytes)
+end
+function Metadata.PrimitiveArray(A::DictEncoding, off::Integer, nbytes::Integer)
+    Metadata.PrimitiveArray(feathertype(eltype(references(A))), Metadata.PLAIN, off, length(A),
+                            nullcount(A), nbytes)
+end
+
+
+writecontents(io::IO, A::Primitive) = writepadded(io, A)
+writecontents(io::IO, A::NullablePrimitive) = writepadded(io, A, bitmask, values)
+writecontents(io::IO, A::List) = writepadded(io, A, offsets, values)
+writecontents(io::IO, A::NullableList) = writepadded(io, A, bitmask, offsets, values)
+writecontents(io::IO, A::BitPrimitive) = writepadded(io, A, values)
+writecontents(io::IO, A::NullableBitPrimitive) = writepadded(io, A, bitmask, values)
+writecontents(io::IO, A::DictEncoding) = writecontents(io, references(A))
+function writecontents(::Type{Metadata.PrimitiveArray}, io::IO, A::ArrowVector)
+    a = position(io)
+    writecontents(io, A)
+    b = position(io)
+    Metadata.PrimitiveArray(A, a, b-a)
+end
+
+
+function writecolumn(io::IO, name::AbstractString, A::ArrowVector{J}) where J
+    vals = writecontents(Metadata.PrimitiveArray, io, A)
+    Metadata.Column(String(name), vals, getmetadata(io, J, A), "")
+end
+function writecolumn(sink::Sink, col::String)
+    writecolumn(sink.io, col, sink.columns[sink.schema[col]])
+end
+writecolumns(sink::Sink) = Metadata.Column[writecolumn(sink, n) for n ∈ Data.header(sink.schema)]
+
+
+function writemetadata(io::IO, ctable::Metadata.CTable)
+    meta = FlatBuffers.build!(ctable)
+    rng = (meta.head+1):length(meta.bytes)
+    writepadded(io, view(meta.bytes, rng))
+    Int32(length(rng))
+end
+writemetadata(sink::Sink) = writemetadata(sink.io, sink.ctable)
+
+
+function Data.close!(sink::Sink)
```

Comment:
this is nice how all the writing happens in one place.

Reply:
Like I said I was contemplating breaking it up a little, but I'm not determined to do it.  I definitely agree that it has to be done in such a way that it does not obscure where in the IO buffer you are at any point in the write process.  Anyway, I'll leave this alone then, I didn't have any particularly inspired ideas about it: I think I just wanted to formalize how we deal with the header and tailer bytes a bit.

Reply:
I like it just like this; before it was a little all over the place in terms of where all the IO happened.